### PR TITLE
Add exploration report on small deals in Filecoin

### PR DESCRIPTION
**Pull request by rvagg**
Details on my exploration of:

1. Installing Lotus and syncing a node
2. Making small deals (attempting to)
3. Filling in many holes in my mental model for how Filecoin works - to this end I've included a section at the top titled "Filecoin storage for the uninitiated" that's a summary of the storage basics that I think might be useful to others (I'm thinking of presenting this in a better form). I would love for @dirkmc to have a quick read through at least that section and correct anything I'm getting wrong.
4. Thoughts on some of the challenges we have in making small deals - much of this is anecdotal and I'm sure more could be added here

### CODE REVIEW

**Comment by rvagg** [comment]
Updated with details of another successful deal! It was waiting for acceptance when I pushed it originally but it's now been accepted and waiting to be sealed.

File: home/rvagg/2021-02-exploration-report-small-filecoin-deals.md [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+
+## Background
+
+### Filecoin storage for the uninitiated
+
+* Typical miners are continually "sealing" new sectors, containing either 32 GiB or 64 GiB of data
+* Sealed sectors have a lifetime, pre-specified by the miner, up to a maximum of 18 months and a minimum of 6 months
+* Sealed sectors contribute to miner "power", power increases a miner's chances of winning block rewards
+* Sectors may, or may not contain data from clients in the form of paid "deals"
+* Sectors not containing deals are called "committed capacity" (CC) sectors, they simply contribute to the total network capacity without storing useful data
```
Comment (ribasushi) [inline home/rvagg/2021-02-exploration-report-small-filecoin-deals.md:11]: At present many sectors will contain some CC portion - due to the fr32 requirements any data-piece must start at an offset that is divisible by the size of that piece. At present there's no "smart packing" of any sort, and also not sufficient amount of incoming deals yet to pack a sector full-full.
This is a reason why miners prefer 16.1G+ fil+ deals - it leaves no CC space at all

File: home/rvagg/2021-02-exploration-report-small-filecoin-deals.md [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+   * CC sectors remove the incentive to self-deal (faux-deal) where miners pretend to have deal data to increase power
+   * Total network capacity, including CC, signals to users that the network has real hardware ready to store data even if those sectors currently don't
+* Sectors that contain deal data are called regular sectors
+* There is no requirement for a mix of data types in sealed sectors (deal / non-deal)
+* Miners have to prove that they retain the entirety of the sealed sectors as-is in regular invervals during each day of the sector's lifetime and are penalised if they can't prove or miss a proving deadline
+* Sectors not proven over a fixed number of consecutive deadlines are terminated and a penalty applied
```
Comment (ribasushi) [inline home/rvagg/2021-02-exploration-report-small-filecoin-deals.md]: You are allowed 1 "free" miss. Then you get penalized every verification. Just like a "daily parking fine", they pile up. After 14 consecutive one - the sector ( and its data ) are burned and you are fined even more / lose all collaterals.

File: home/rvagg/2021-02-exploration-report-small-filecoin-deals.md [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+* Clients who have deals in terminated sectors have unpaid funds returned to them
```
Comment (ribasushi) [inline home/rvagg/2021-02-exploration-report-small-filecoin-deals.md]: This is... complicated. I believe the protocol does not do 100% refunds, to disincentivise a miner self-dealing and "exitting cleanly". The spec is murky on this, I just have a vague recollection it was not "you just get a full refund". Please check with CE/RT

File: home/rvagg/2021-02-exploration-report-small-filecoin-deals.md [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+* Typical miners are continually "sealing" new sectors, containing either 32 GiB or 64 GiB of data
+* Sealed sectors have a lifetime, pre-specified by the miner, up to a maximum of 18 months and a minimum of 6 months
+* Sealed sectors contribute to miner "power", power increases a miner's chances of winning block rewards
+* Sectors may, or may not contain data from clients in the form of paid "deals"
+* Sectors not containing deals are called "committed capacity" (CC) sectors, they simply contribute to the total network capacity without storing useful data
+   * CC sectors remove the incentive to self-deal (faux-deal) where miners pretend to have deal data to increase power
+   * Total network capacity, including CC, signals to users that the network has real hardware ready to store data even if those sectors currently don't
+* Sectors that contain deal data are called regular sectors
+* There is no requirement for a mix of data types in sealed sectors (deal / non-deal)
+* Miners have to prove that they retain the entirety of the sealed sectors as-is in regular invervals during each day of the sector's lifetime and are penalised if they can't prove or miss a proving deadline
```
Comment (ribasushi) [inline home/rvagg/2021-02-exploration-report-small-filecoin-deals.md]: The "regular interval" is generally correct, *BUT* there are "drills" once a week or something where you get asked to produce a WPoST out of the blue at a random time. Not sure what the parameters are, just observed it way back on testnet.

File: home/rvagg/2021-02-exploration-report-small-filecoin-deals.md [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+* Sectors not proven over a fixed number of consecutive deadlines are terminated and a penalty applied
+* Clients who have deals in terminated sectors have unpaid funds returned to them
+* Miners lock up collateral when they seal sectors, which is returned when the sector expires without fault, as an incentive to keep them active during the sector lifetime
```
Comment (ribasushi) [inline home/rvagg/2021-02-exploration-report-small-filecoin-deals.md:20]: The size of the collateral is comprised of 3 parts, 2 determined by network state and the last one chosen by the client ( this data is *that* important to me )

File: home/rvagg/2021-02-exploration-report-small-filecoin-deals.md [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+* Deals can be made in the live market by a client contacting a miner directly and:
+    * Requesting deal conditions (price per GiB per epoch (30 seconds), min size and max size)
+    * Proposing a specific deal
+    * Having a deal accepted by a miner
+    * Transferring the data to the miner
+    * The miner including the deal data into a sector they are committing (currently or queued)
+    * Having deal confirmation placed on the chain pointing to its location in the sealed sector
```
Comment (ribasushi) [inline home/rvagg/2021-02-exploration-report-small-filecoin-deals.md]: One more step after that - sealing the sector itself and placing the replica proof on chain .

File: home/rvagg/2021-02-exploration-report-small-filecoin-deals.md [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+* Deals can also be made "offline" where the deal is made but the data is shipped (perhaps manually) to the miner who loads it into a sector
```
Comment (ribasushi) [inline home/rvagg/2021-02-exploration-report-small-filecoin-deals.md:30]: Critically as a client you still need access to the data. This stems from the deal proposal being client-driven and immutable in the current design ( the payment channels are tied to the content-address of the proposal )

Since you as a client are the only one who can send a new signed proposal, you must include the piececid/commp in the deal, therefore must be in possession of the data at some point ( or have the miner give you commp and trust they will not screw you over )

File: home/rvagg/2021-02-exploration-report-small-filecoin-deals.md [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+* Deal funds pay out to the miner over the course of the lifetime of the deal to ensure they continue proving retention of data
```
Comment (ribasushi) [inline home/rvagg/2021-02-exploration-report-small-filecoin-deals.md:31]: The vesting schedule looks to not be entirely linear here either, not sure if important.

File: home/rvagg/2021-02-exploration-report-small-filecoin-deals.md [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+* Deals can either be regular deals or "verified deals"
```
Comment (ribasushi) [inline home/rvagg/2021-02-exploration-report-small-filecoin-deals.md]: Try not to use "verified" in coms going forward, it's now all "Fil+" / "Filecoin Plus"

File: home/rvagg/2021-02-exploration-report-small-filecoin-deals.md [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+    * Verified deals come from an additional social layer outside of Filecoin, called Filecoin Plus (FIL+)
+    * Notaries are organisations that can hand out "DataCap" to clients
+    * Notaries are selected by the Filecoin community, distributed around the world
+    * "DataCap" is a quota that can be spent, along with FIL, to make verified deals
+    * Verified deals contribute 10x to miner power, proportional to how much space they occupy in a sector
```
Comment (ribasushi) [inline home/rvagg/2021-02-exploration-report-small-filecoin-deals.md]: Notably, since from a miner perspective this is "10x the data", the provider collaterals are multiplied by 10 as well ( was a surprise to a number of miners )

File: home/rvagg/2021-02-exploration-report-small-filecoin-deals.md [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+    * Therefore, verified deals should be attractive to miners who may charge less FIL to get them
```
Comment (ribasushi) [inline home/rvagg/2021-02-exploration-report-small-filecoin-deals.md]: On the various explorers you will see "Raw Power" vs "Quality Adjusted Power" discrepancy - this is the effect of fil+, e.g.: https://filfox.info/en/address/f010479

File: home/rvagg/2021-02-exploration-report-small-filecoin-deals.md [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+
+### Deal incentives & challenges
+
+* Miners bear responsibility for publishing deals to the chain, with high gas costs this can be in the order of $5
+* Recent changes allow for batch publishing deals, but anecdotally this hasn't changed the economics for miners much (need more context here - is it because the batch publish is still quite expensive, or because there are not enough deals to batch?)
```
Comment (ribasushi) [inline home/rvagg/2021-02-exploration-report-small-filecoin-deals.md]: It seems that because of the amount of disparate state structures to update, the batching saving is not linear / leads to hard to predict message gas parameters. Miners are experimenting trying to get a handle on this - the feature is way too new.

File: home/rvagg/2021-02-exploration-report-small-filecoin-deals.md [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+* Accepting deals can offset the cost of sealing and proving, but this biases toward larger deals
+* Verified deals have the 10x increase in power, but they are scarce because FIL+ is so new, currently unknown the scale of incentive impact for miners for large and small deals
+* Bugs in Lotus have made miners gun-shy on doing anything more than CC:
+  * Transferring deal data is buggy and transfers often fail (anecdotal)
+  * Retrieval is buggy and can cause crashes, which are problematic when you need 100% uptime to prove sectors (anecdotal)
+* A lot of large miners don't seem to have transitioned from CC sealing to accepting deals - could be immaturity of market
+* Many miners don't respond to `query-ask` at all. Of those that do, the majority appear to reject deals even when those deals fit within the ask parameters. So the `query-ask` mechanism seems of limited use in its current form.
```
Comment (ribasushi) [inline home/rvagg/2021-02-exploration-report-small-filecoin-deals.md:55]: Most critically: `query-ask` does not interface with https://docs.filecoin.io/mine/lotus/miner-configuration/#using-filters-for-fine-grained-storage-and-retrieval-deal-acceptance

That is - query-ask is likely to tell you "All are welcome, free!"
And then the filter will tell you "no no, you are not on my list"

File: home/rvagg/2021-02-exploration-report-small-filecoin-deals.md [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+
+Filecoin currently only stores ~5 Pb of deal data and growth rate is not large.
+
+## Installing and Running Lotus
+
+Followed instructions at https://docs.filecoin.io/get-started/lotus/installation/ for an AWS instance
+
+ * Large AMD EPYC instance, Ubuntu 20.04
+ * Installed dependencies (used Go Snap rather than Go tarball)
+ * Cloned and build Lotus with CPU SHA support
```
Comment (ribasushi) [inline home/rvagg/2021-02-exploration-report-small-filecoin-deals.md]: You actually do not need this ( nor an AMD ) for just validating the chain. The SHA-NI helps exclusively with mining.

File: home/rvagg/2021-02-exploration-report-small-filecoin-deals.md [hunk 15]
```diff
@@ -0,0 +1,134 @@
...
+ * Used minimal stateroots snapshot @ https://fil-chain-snapshots-fallback.s3.amazonaws.com/mainnet/minimal_finality_stateroots_latest.car
+ * Started manually with `lotus daemon --import-snapshot`
+ * Stopped, then started via systemd, let it sync, watching the log output with `tail -f`
+ * Can interact with the daemon via the CLI with `lotus client` and `lotus wallet`, etc. as the `lotus` user (appears to be user-restricted IPC)
+ * Made a wallet and transferred some FIL in it for experimenting
+
+Interesting commands:
+
+ * `lotus sync wait` will run until the chain is synced, zero exit status on success
+ * `lotus wallet` - `lotus wallet new`, `lotus wallet new`, `lotus wallet balance`, `lotus wallet export` (export privkey for backup)
```
Comment (ribasushi) [inline home/rvagg/2021-02-exploration-report-small-filecoin-deals.md:77]: Using local ledger with a cloud node is also a thing ( docs are lacking / pending )
https://hackmd.io/iZJacRQmSN-obAjEypEanw

File: home/rvagg/2021-02-exploration-report-small-filecoin-deals.md [hunk 16]
```diff
@@ -0,0 +1,134 @@
...
+ * `lotus client`
+   * `lotus client import <path>` to import file(s) into the client (somewhere?), presumably as UnixFS (?), gets a CID
+   * `lotus client local` shows you what you have imported and their CIDs
+   * `lotus client balances` shows you funds that you have active in the "market" (where you have to have funds in order to make deals etc. i.e. out of your wallet)
+   * `lotus client find <CID>` lets you query the network for a CID
```
Comment (ribasushi) [inline home/rvagg/2021-02-exploration-report-small-filecoin-deals.md]: It does not work :(  The "query network" implementation is merely a stub, to be written:
https://github.com/filecoin-project/go-fil-markets/blob/v1.1.9/discovery/impl/discovery.go#L7-L9

Currently it *appears* to work because it is querying a list of peers you already dealt with previously: https://github.com/filecoin-project/go-fil-markets/blob/v1.1.9/storagemarket/impl/client.go#L404-L410

File: home/rvagg/2021-02-exploration-report-small-filecoin-deals.md [hunk 17]
```diff
@@ -0,0 +1,134 @@
...
+
+Deals:
+
+   * `lotus client query-ask <minerID>` ask a miner for their deal terms - returns FIL asking price per GiB per epoch (30 seconds) for both regular and verified deals, along with minimum and maximum deal sizes.
```
Comment (ribasushi) [inline home/rvagg/2021-02-exploration-report-small-filecoin-deals.md]: See my note about custom filters above

File: home/rvagg/2021-02-exploration-report-small-filecoin-deals.md [hunk 18]
```diff
@@ -0,0 +1,134 @@
...
+   * `lotus client list-asks` will do a scan of the network for miners with power and do a `query-ask` on each of them, reporting the output to stdout, one miner per line (~450)
+   * `lotus client deal` without additional arguments takes you through a prompted deal-making process that asks for your data CID, a miner ID, whether it's a verified deal; it queries the miner, asks you to confirm the asking price and starts the deal process if you say `yes`.
+   * `lotus client list-deals` to show you deals in progress and their status (Note: the status identifiers are not self-explainatory, the best source of documentation seems to be the markets code: https://github.com/filecoin-project/go-fil-markets/blob/master/storagemarket/dealstatus.go)
+   * `lotus client deal-stats` seems to be the best way to view the outcome of your deals once they disappear off `list-deals`
```
Comment (ribasushi) [inline home/rvagg/2021-02-exploration-report-small-filecoin-deals.md]: > once they disappear 

There's the oddly named `--show-failed` flag that "brings them back"
There's also `-v` to that command

File: home/rvagg/2021-02-exploration-report-small-filecoin-deals.md [hunk 19]
```diff
@@ -0,0 +1,134 @@
...
+   * The Lotus daemon log also has output related to deals, you can grep for the status, e.g. `grep StorageDealError ~/var/lotus-daemon.log` (mostly just `"error reading Response message: EOF"`)
+
+### Making deals
+
+* Imported a 9.6 Mb (`10016091` b) JPG file to Lotus with `lotus client import`. Some notes about sizes:
+  * Note that data is _padded_ when making deals. First 2 empty bits per 254 bytes are added (called "fr32 padding", `Math.ceil(10016091 * (256/254))`), then it's rounded up to the next nearest power-of-two value as "pieces" stored in Filecoin have to fit neatly into a proving tree (`2**Math.ceil(Math.log2(Math.ceil(10016091*(256/254))))`).
```
Comment (ribasushi) [inline home/rvagg/2021-02-exploration-report-small-filecoin-deals.md:97]: There's a wolfram snippet circulating in the community for this: https://www.wolframalpha.com/input/?i=x+%3D+10016091%3B+128*%28+2%5E%28+ceil%28+log2%28+ceil+%28+x+%2F127+%29+%29+%29+%29+%29

File: home/rvagg/2021-02-exploration-report-small-filecoin-deals.md [hunk 20]
```diff
@@ -0,0 +1,134 @@
...
+  * The additional padding is done transparently, but it leads to wonky numbers in the deal process that aren't explained well (and online documentation is inconsistent on this and doesn't explain it well - TODO some PRs)
```
Comment (ribasushi) [inline home/rvagg/2021-02-exploration-report-small-filecoin-deals.md:98]: Worse - offline deal interface takes the unpadded values, so users need to juggle "do I multiply by 128 or by 127...?"
https://docs.filecoin.io/store/lotus/very-large-files/#calculate-the-datasize

Difficult to fix because all of this is now entrenched API.

File: home/rvagg/2021-02-exploration-report-small-filecoin-deals.md [hunk 21]
```diff
@@ -0,0 +1,134 @@
...
+  * The most consistent number shown is power-of-two _minus_ the fr32 padding. (`2**Math.ceil(Math.log2(Math.ceil(10016091*(256/254))))/(256/254)/1024/1024` as `MiB`, or `15.875`, presented as `15.88 MiB` by Lotus - clear as mud).
+  * Aside: an implication of padding is that when packing data in for making efficient deals, you should get as close to the "power-of-two - fr32 padding" size as possible. The larger the data, the more waste you'll have. We had to do this for Filecoin Discover data to edge up as close as possible to 1 GiB pieces.
```
Comment (ribasushi) [inline home/rvagg/2021-02-exploration-report-small-filecoin-deals.md]: The inverse of this implication is that certain miner/client combos with access to datacap make deals that pad to precisely 16.001GiB, to enjoy a full 10x sector but effectively "lock" just half the data they have available.

File: home/rvagg/2021-02-exploration-report-small-filecoin-deals.md [hunk 22]
```diff
@@ -0,0 +1,134 @@
...
+* Found some miners
+  * Used the MinerX list (link to spreadsheet omitted)
+  * Ran `query-ask` against all of them (`for i in $(cat minerx-fellows.txt); do echo "querying $i"; lotus client query-ask $i >> minerx-fellows.txt.query-ask; done`); got 34 out of 36 responses.
```
Comment (ribasushi) [inline home/rvagg/2021-02-exploration-report-small-filecoin-deals.md:104]: > got 34 out of 36 responses.

Uh oh...

File: home/rvagg/2021-02-exploration-report-small-filecoin-deals.md [hunk 23]
```diff
@@ -0,0 +1,134 @@
...
+  * Slate did some work on finding miners (not sure how up to date this is), collated @ https://sentinel.slate.host/api/mapped-static-global-miners and includes some miners that accept small deals too.
+* Ran `lotus client deal` to step through some deals interactively with miners I'd narrowed down. Got the following results for my file and submitted deals with most of these:
+  - f019104 ~0.0093312 FIL (0.000000018 FIL per epoch) ~$0.32
+  - f019041 ~0.0002592 FIL (0.0000000005 FIL per epoch) ~$0.01
+  - f066596 ~0.010368 FIL (0.00000002 FIL per epoch) ~$0.36
+  - f03134 ~8100 FIL (0.015625 FIL per epoch) ~$283,500 _(obviously I said `no` to this one)_
+  - f023565 ~0.0000000000005184 FIL (0.000000000000000001 FIL per epoch) $0
+  - f022352 ~0.0001458 FIL (0.00000000028125 FIL per epoch) ~$0.01 (0.005103)
+* Deal process is very slow and involves:
+  * Moving funds from the local wallet into the "StorageMarket" actor, to setup for a kind of escrow so the miner knows they're going to get paid
```
Comment (ribasushi) [inline home/rvagg/2021-02-exploration-report-small-filecoin-deals.md:115]: This is additionally hampered by https://github.com/filecoin-project/lotus/issues/5543#issuecomment-775339608

File: home/rvagg/2021-02-exploration-report-small-filecoin-deals.md [hunk 24]
```diff
@@ -0,0 +1,134 @@
...
+  * The miner deciding whether to accept the funds. Miners set default deal parameters but can set additional configuration options and even insert an "deal acceptance script" to their deal flow to accept and reject deals based on all of the parameters. See https://docs.filecoin.io/mine/lotus/miner-configuration/#using-filters-for-fine-grained-storage-and-retrieval-deal-acceptance
+    * This can be slow too, I had deals stuck in `StorageDealCheckForAcceptance` for extended periods, I'm not sure what this is but maybe it's waiting for manual acceptance by the miner?
```
Comment (ribasushi) [inline home/rvagg/2021-02-exploration-report-small-filecoin-deals.md:117]: Usually it is seeing the channel land on chain - see above. You get an idea by running `lotus mpool pending --local` - if it returns a result you are in trouble

File: home/rvagg/2021-02-exploration-report-small-filecoin-deals.md [hunk 25]
```diff
@@ -0,0 +1,134 @@
...
+  * Transferring the data via graphsync to the miner
+  * The miner setting up the deal with the chain (does this come before or after the next step?)
+  * The miner queueing the piece for sealing (can be slow, miners can choose to batch deals, which defer deals until enough have been selected, and can set maximum number of deals per sector too)
```
Comment (ribasushi) [inline home/rvagg/2021-02-exploration-report-small-filecoin-deals.md:120]: Generally allowing 12~24h for queue+sealing is reasonable here. Anything under 12h is "WHOA!!! Really?!"

**Review by ribasushi — APPROVED** [review]
This is an excellent report. I left a ton of comments - all are of the "but wait, there's more" variety.

Feel free to incorporate or leave out anything and everything at your discretion.

File: home/rvagg/2021-02-exploration-report-small-filecoin-deals.md (hunk 1 above)
Comment (rvagg) [reply home/rvagg/2021-02-exploration-report-small-filecoin-deals.md:11]: > This is a reason why miners prefer 16.1G+ fil+ deals - it leaves no CC space at all

from what I understand, deals _must_ be ^2 sized (i.e. the client has to make a deal ^2 sized, they can't propose a 16.1 deal), so this would be 16.1G user data but a 32G deal wouldn't it? So the preference here would simply come from the fact that they are able to fill up a sector with a deal while only transferring the minimum amount of data from the client? is the data transfer the only benefit here?

File: home/rvagg/2021-02-exploration-report-small-filecoin-deals.md (hunk 19 above)
Comment (rvagg) [reply home/rvagg/2021-02-exploration-report-small-filecoin-deals.md:97]: yeah, saw it, and I hate it, so I'm not mentioning it

File: home/rvagg/2021-02-exploration-report-small-filecoin-deals.md (hunk 21 above)
Comment (rvagg) [reply home/rvagg/2021-02-exploration-report-small-filecoin-deals.md]: this is back to my question above - "deals that pad to precisely 16.001GiB" - I was under the impression that "deal size" that you pay for is the entire thing, but apparently it's not? you only pay for pre-padded data?

File: home/rvagg/2021-02-exploration-report-small-filecoin-deals.md (hunk 1 above)
Comment (ribasushi) [reply home/rvagg/2021-02-exploration-report-small-filecoin-deals.md:11]: >  the fact that they are able to fill up a sector with a deal while only transferring the minimum amount of data from the client? 

given how difficult it is currently - that's a *very* big deal ;)

File: home/rvagg/2021-02-exploration-report-small-filecoin-deals.md (hunk 21 above)
Comment (ribasushi) [reply home/rvagg/2021-02-exploration-report-small-filecoin-deals.md]: No, you are supposed to pay for the entire thing. Transfer savings is what makes this attractive.

Comment (rvagg) [reply home/rvagg/2021-02-exploration-report-small-filecoin-deals.md]: got it, that's really interesting context, I thought that the padded content gets transferred but if not then that's a nice efficiency gain on data transfer

File: home/rvagg/2021-02-exploration-report-small-filecoin-deals.md (hunk 19 above)
Comment (rvagg) [reply home/rvagg/2021-02-exploration-report-small-filecoin-deals.md:97]: fyi there's a further conversation related to this piece of the documentation: https://github.com/filecoin-project/filecoin-docs/issues/705

File: home/rvagg/2021-02-exploration-report-small-filecoin-deals.md (hunk 21 above)
Comment (ribasushi) [reply home/rvagg/2021-02-exploration-report-small-filecoin-deals.md]: The only thing transferred is the indiviual blocks of the DAG, selector-style. `car` files themselves never come into the equation, let alone the padding on top of them.
Remember this entire thing being a... thing: https://github.com/filecoin-project/go-fil-markets/issues/365
Storage is "mandatory-graphsync" in exactly the same way ( currently ).

Comment (rvagg) [reply home/rvagg/2021-02-exploration-report-small-filecoin-deals.md]: yeah, I do remember now, all the gory graphsync stuff comes into focus in this point doesn't it? makes sense but is absolutely obscured by Lotus always talking about this "datasize" thing

**Comment by rvagg** [comment]
This has been a very fruitful exercise. I've now updated the doc with additional notes from feedback I received here and elsewhere, along with some final details about deal status, copied below:

### Deal outcomes

Successfully got 3 deals into `StorageDealActive` as of final doc edit. This was for a 9.6 Mb file, that turned into a 15.88 MiB padded deal size to consume 16 MiB in the miners' sectors.

* The prices of the deals are `0.00014706421875`, `0` (FIL+ with same miner) and `0.00081599375` (FIL+ with different miner), total of `0.00096305796875`, or $0.0337 at current FIL price.
* `lotus client balances` shows   `Escrowed Funds: 0.0009616335625 FIL` and `Locked Funds: 0.0009616335625 FIL` - am assuming these are the same thing at this stage of the deal (they were different during deal processing).
* Unsure where the `0.00000142440625` difference between total deal amount and locked funds amount is - assuming it's an amount already paid out to miners as of the time of writing.
* The cost to me for making these 3 deals in total was `0.025837882018354755` FIL, or $0.90 at current FIL price. This was mostly consumed getting data into the StorageMarket actor to make the deals.
* The cost to the miners for making these deals is unaccounted for (it might be an interesting exercise to dig into this and account for their spend and compare it to their earn because there's going to be quite a difference).
* Question: If you are mostly making deals, can you just transfer a bulk amount into the StorageMarket actor from your wallet so you pay for the transfer once and use it up across many deals?

File: home/rvagg/2021-02-exploration-report-small-filecoin-deals.md [hunk 26]
```diff
@@ -54,7 +63,7 @@ Followed instructions at https://docs.filecoin.io/get-started/lotus/installation
 
  * Large AMD EPYC instance, Ubuntu 20.04
  * Installed dependencies (used Go Snap rather than Go tarball)
- * Cloned and build Lotus with CPU SHA support
+ * Cloned and build Lotus with CPU SHA support (learned later that these flags don't impact `lotus` compile, the SHA extnsions are only used for mining)
```
Comment (ribasushi) [inline home/rvagg/2021-02-exploration-report-small-filecoin-deals.md:66]: Now that I read this I realize I mislead you *a little*. Yes, SHA acceleration has no effect on actual chainsync. But it does help noticeably when calculating the CommP to shove into the deal proposal ( underneath all layers it's still sha2-256 ). So a user making a ton of deals could potentially benefit, especially if they are not transfering the data ( offline deals of some capacity )

File: home/rvagg/2021-02-exploration-report-small-filecoin-deals.md [hunk 27]
```diff
@@ -132,3 +144,15 @@ Deals:
...
+* `lotus client balances` shows   `Escrowed Funds: 0.0009616335625 FIL` and `Locked Funds: 0.0009616335625 FIL` - am assuming these are the same thing at this stage of the deal (they were different during deal processing).
+* Unsure where the `0.00000142440625` difference between total deal amount and locked funds amount is - assuming it's an amount already paid out to miners as of the time of writing.
+* The cost to me for making these 3 deals in total was `0.025837882018354755` FIL, or $0.90 at current FIL price. This was mostly consumed getting data into the StorageMarket actor to make the deals.
+* The cost to the miners for making these deals is unaccounted for (it might be an interesting exercise to dig into this and account for their spend and compare it to their earn because there's going to be quite a difference).
+* Question: If you are mostly making deals, can you just transfer a bulk amount into the StorageMarket actor from your wallet so you pay for the transfer once and use it up across many deals?
```
Comment (ribasushi) [inline home/rvagg/2021-02-exploration-report-small-filecoin-deals.md:157]: ☝️ that's definitely a @dirkmc question

Comment (dirkmc) [reply home/rvagg/2021-02-exploration-report-small-filecoin-deals.md:157]: You can use `lotus wallet market add`:
```
$ lotus wallet market add --help

NAME:
  lotus wallet market add - Add funds to the Storage Market Actor
USAGE:
  lotus wallet market add [command options] <amount>
OPTIONS:
  --from value, -f value   Specify address to move funds from, otherwise it will use the default wallet address
  --address value, -a value Market address to move funds to (account or miner actor address, defaults to --from address)
  --help, -h         show help (default: false)
```

Comment (rvagg) [reply home/rvagg/2021-02-exploration-report-small-filecoin-deals.md:157]: Thanks @dirkmc - so if I use this and transfer a whole lot of FIL into there, will I incur any further charges in making deals or do those all fall on the miner now? I think that covers the `AddBalance` transfers here, but what about the `Send` and `Exec`?

Screenshot 2021-03-01 at 9 34 40 pm

Comment (dirkmc) [reply home/rvagg/2021-02-exploration-report-small-filecoin-deals.md:157]: AFAIK the only message the client needs to send is AddBalance to put the funds required for the deal in escrow with the Storage Market Actor.
Send is for sending FIL to an address, and Exec I believe is when you create a miner. So those messages may not be related to deals your actor has been making.

File: home/rvagg/2021-02-exploration-report-small-filecoin-deals.md [hunk 28]
```diff
@@ -0,0 +1,158 @@
...
+
+* Typical miners are continually "sealing" new sectors, containing either 32 GiB or 64 GiB of data
+* Sealed sectors have a lifetime, pre-specified by the miner, up to a maximum of 18 months and a minimum of 6 months
+* Sealed sectors contribute to miner "power", power increases a miner's chances of winning block rewards
+* Sectors may, or may not contain data from clients in the form of paid "deals"
+* Sectors not containing deals are called "committed capacity" (CC) sectors, they simply contribute to the total network capacity without storing useful data
+   * CC sectors remove the incentive to self-deal (faux-deal) where miners pretend to have deal data to increase power
+   * Total network capacity, including CC, signals to users that the network has real hardware ready to store data even if those sectors currently don't
+* Sectors that contain deal data are called "regular sectors"
+* There is no requirement for a mix of data types in sealed sectors (deal / non-deal)
```
Comment (alanshaw) [inline home/rvagg/2021-02-exploration-report-small-filecoin-deals.md:15]: I found this hard to parse in my head but I think it means the data type in a sealed sector _can_ be mixed deal/non-deal?

File: home/rvagg/2021-02-exploration-report-small-filecoin-deals.md [hunk 29]
```diff
@@ -0,0 +1,158 @@
...
+  - f066596 0 FIL $0
+  - f03134 ~8100 FIL (0.015625 FIL per epoch) ~$283,500 _(obviously I said `no` to this one)_
+  - f023565 0 FIL $0
+  - f022352 0 FIL $0
+  - f071624 ~0.000000405 FIL (0.00000000000078125 FIL per epoch)
+  - f0135078 ~0.000000405 FIL (0.00000000000078125 FIL per epoch)
+  - f058369 ~0.00081 FIL (0.0000000015625 FIL per epoch)
+  - f010479 ~0 FIL (0 FIL per epoch)
+  - f01247 ~0 FIL (0 FIL per epoch)
+* Successfully made a deal with the same miner as for my unverified deal, [`f022352`](https://filfox.info/en/address/f022352). As of writing, still waiting for PreCommit, ~5 hours later. 0 FIL instead of 0.0001458 FIL.
```
Comment (alanshaw) [inline home/rvagg/2021-02-exploration-report-small-filecoin-deals.md:145]: Interesting, does the retrieval cost change at all between these two deals?

Comment (alanshaw) [reply home/rvagg/2021-02-exploration-report-small-filecoin-deals.md:145]: ...and I'm also interested if you are actually able to retrieve your data from the miner?

File: home/rvagg/2021-02-exploration-report-small-filecoin-deals.md (hunk 28 above)
Comment (rvagg) [reply home/rvagg/2021-02-exploration-report-small-filecoin-deals.md:15]: Poorly worded, sorry. So this image under "sector quality" in the specs is kind of interesting to illustrate what this is talking about (even though this image is designed to talk about "quality" -> "power", hence the multipliers):

sector quality

Basically, a sector can contain no deals, some deals, all deals, and those deals can be regular and verified, varying sizes, in any mix, as long as they fit nicely into the power-of-two offsets.

File: home/rvagg/2021-02-exploration-report-small-filecoin-deals.md (hunk 29 above)
Comment (rvagg) [reply home/rvagg/2021-02-exploration-report-small-filecoin-deals.md:145]: ... no ... I haven't poked at retrieval at all in this process so I'm pretty ignorant, and I don't expect that to be straightforward, which is why I've stopped sort to focus on the _getting things in_ question.
